docs(adr): revise ADR 017 subspace fee grant (#1070)
desmos-labs/desmos: docs/architecture/adr-017-subspace-fee-grant.md

The `x/feegrant` module already provides an interface (`FeeAllowanceI`) to represent a generic allowance, along with a set of useful implementations like `BasicAllowance`, `AllowedMsgAllowance` and `PeriodicAllowance`. Since we are not going to introduce new kinds of allowances, nor we are going to edit how an allowance is represented, we will reuse these types.

#### Grant

> In order to represent a fee grant granted to a user or user group, we will implement a custom `Grant` type. This will contain the `granter`, `grantee`, which is the interface to include user and user group, and what kind of `allowance` is granted to the grantee:

Reviewer: ```suggestion
In order to represent a fee grant granted to a user or user group, we will implement a custom `Grant` type. This will contain a `granter` field, a `grantee` (an interface to represent either a user or user group), and what kind of `allowance` is granted to the grantee:
```

  option (gogoproto.equal) = true;
  option (gogoproto.goproto_stringer) = true;

  uint32 group_id = 1 [
    (gogoproto.customname) = "GroupID",
    (gogoproto.moretags) = "yaml:\"group_id\""
  ];
}
```

### Store

#### User grant

Each subspace fee granted allowance will be stored using keys with the following structure:
```
SubspaceUserAllowancePrefix | SubspaceID | GranteeAddress |-> Protobuf(Grant)
```

This structure allows subspace owner/admins to easily manage their subspace grants by iterating over all grants with the given subspace id, which will be the most used query. On the other hand, it allows the following queries:
1. retrieving a specific grant.  
   This will require grantees to know who the granter address, and can be done in O(1).
2. retrieving all the grants that a user has been granted. 
   This can be done by iterating over all the subspace's grants in O(N), where N is the number of all the grants inside the given subspace.

#### Group grant

In order to simplify the management of group allowances, each granted group allowance will be stored using the following key:
```
SubspaceGroupAllowancePrefix | SubspaceID | GroupID | -> Protobuf(Grant)

> ```

Author: Although `UserGrant` state can be taken care by `x/authz`, I still decided to store the state in `x/subspaces` since it is in a particular way highly depending `x/subspaces`. It also avoid the changes of `x/authz` keeper influencing `x/subspaces`.
